### PR TITLE
Story 3

### PR DESCRIPTION
**Pull request by melody-sylvestre**

### CODE REVIEW

File: src/Entities/Project.php [hunk 1]
```diff
@@ -11,14 +13,32 @@ class Project
...
+        $this->client_name = $client_name;
+        $this->client_logo = $client_logo;
+        $this->users = $users;
+    }
+
+    private function validateUsersArray(array $users)
+    {
+        foreach ($users as $user) {
+            if (!($user instanceof User)) {
+                throw new InvalidUserArrayDatatypeException("Invalid user array\n");
```
Comment (norayna) [inline src/Entities/Project.php]: this error message could be more informative, "Invalid user array" doesn't necessarily explain what happened very clearly

File: src/Exceptions/InvalidUserArrayDatatypeException.php [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace ProjectManagementApi\Exceptions;
+
+class InvalidUserArrayDatatypeException extends \Exception
```
Comment (norayna) [inline src/Exceptions/InvalidUserArrayDatatypeException.php]: you should 'use Exception' at the top of the file, similar to your other exception class

File: src/Entities/Project.php (hunk 1 above)
Comment (melody-sylvestre) [reply src/Entities/Project.php]: done

File: src/Exceptions/InvalidUserArrayDatatypeException.php (hunk 2 above)
Comment (melody-sylvestre) [reply src/Exceptions/InvalidUserArrayDatatypeException.php]: done

File: src/Hydrators/UserHydrator.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+namespace ProjectManagementApi\Hydrators;
+
+use ProjectManagementApi\DatabaseConnection;
+use ProjectManagementApi\Entities\User;
+
+class UserHydrator
+{
+    public static function getUsersByProjectId(DatabaseConnection $db, int $id): array
+    {
+        $query = $db->prepare(
```
Comment (norayna) [inline src/Hydrators/UserHydrator.php:12]: i don't quite understand how you've used AND here, i thought it would need to be WHERE instead?

Comment (DanAbsolon) [reply src/Hydrators/UserHydrator.php:12]: done

File: src/Tests/Entities/ProjectTest.php [hunk 4]
```diff
@@ -78,7 +79,41 @@ public function testSuccessToAssociativeArray()
             'deadline' => '21/12/2012',
             'overdue' => true
         ];
-        $actualOutput = $testProject->toAssociativeArray();
+        $actualOutput = $testProject->toAssociativeArrayFewerProperties();
```
Comment (norayna) [inline src/Tests/Entities/ProjectTest.php:90]: your test method is still called `testSuccessToAssociativeArray`

Comment (melody-sylvestre) [reply src/Tests/Entities/ProjectTest.php:90]: done

File: src/Tests/Entities/ProjectTest.php [hunk 5]
```diff
@@ -78,7 +79,41 @@ public function testSuccessToAssociativeArray()
...
         $this->assertEquals($expectedOutput, $actualOutput);
     }
+
+    public function testSuccessToAssociativeArrayAllProperties()
+    {
+        $testUsers = [];
+        $testUserOne = new User(1, 'name', 'avatar', 'role');
+        $testUserTwo = new User(2, 'name', 'avatar', 'role');
+        $testUserThree = new User(3, 'name', 'avatar', 'role');
```
Comment (norayna) [inline src/Tests/Entities/ProjectTest.php]: you need to mock these users

File: src/Tests/Entities/ProjectTest.php [hunk 6]
```diff
@@ -78,7 +79,41 @@ public function testSuccessToAssociativeArray()
...
+                            'users' => [
+                                ['id' => 1, 'name' => 'name', 'avatar' => 'avatar', 'role' => 'role'],
+                                ['id' => 2, 'name' => 'name', 'avatar' => 'avatar', 'role' => 'role'],
+                                ['id' => 3, 'name' => 'name', 'avatar' => 'avatar', 'role' => 'role']]
+                            ];
+        $actualOutput = $testProject->toAssociativeArrayAllProperties();
+        $this->assertEquals($expectedOutput, $actualOutput);
+    }
+
+    public function testFailureConstructor_IncorrectUserArrayDataType()
```
Comment (norayna) [inline src/Tests/Entities/ProjectTest.php]: this method should be moved to be with the other constructor tests

File: src/Tests/Entities/ProjectTest.php [hunk 7]
```diff
@@ -78,7 +79,41 @@ public function testSuccessToAssociativeArray()
...
+                            'client_id' => '1', 
+                            'deadline' => '21/12/2012', 
+                            'overdue' => true,
+                            'client_name' => 'client name',
+                            'client_logo' => 'http://dummyimage.com/200x200.png/ff4444/ffffff',
+                            'users' => [
+                                ['id' => 1, 'name' => 'name', 'avatar' => 'avatar', 'role' => 'role'],
+                                ['id' => 2, 'name' => 'name', 'avatar' => 'avatar', 'role' => 'role'],
+                                ['id' => 3, 'name' => 'name', 'avatar' => 'avatar', 'role' => 'role']]
+                            ];
```
Comment (norayna) [inline src/Tests/Entities/ProjectTest.php]: your indents here are too indenty

File: src/Tests/Entities/ProjectTest.php (hunk 5 above)
Comment (melody-sylvestre) [reply src/Tests/Entities/ProjectTest.php]: done

Comment (robin-hodge) [reply src/Tests/Entities/ProjectTest.php]: done!

File: src/Tests/Entities/ProjectTest.php (hunk 6 above)
Comment (melody-sylvestre) [reply src/Tests/Entities/ProjectTest.php]: done

Comment (robin-hodge) [reply src/Tests/Entities/ProjectTest.php]: done!

File: src/Tests/Entities/ProjectTest.php (hunk 7 above)
Comment (robin-hodge) [reply src/Tests/Entities/ProjectTest.php]: done!

Comment (melody-sylvestre) [reply src/Tests/Entities/ProjectTest.php]: done

File: project.php [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+use ProjectManagementApi\Hydrators\ProjectHydrator;
+use ProjectManagementApi\Response;
+
+header('Access-Control-Allow-Origin: *');
+header('Content-Type: application/json');
+
+try {
+    $db = new DatabaseConnection('mysql:host=db; dbname=project_manager', 'root', 'password');
+
+    try {
```
Comment (norayna) [inline project.php:16]: try inside of try?

Comment (melody-sylvestre) [reply project.php:16]: OK

Comment (robin-hodge) [reply project.php:16]: OK

File: project.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        $project_id = $_GET['id'];
+
+        $project = ProjectHydrator::getProjectById($db, $project_id);
+        $projectAsAssociativeArray = $project->toAssociativeArrayAllProperties();
+
+        $response = new Response('Successfully retrieved project', $projectAsAssociativeArray);
+        http_response_code(200);
+        echo json_encode($response);
+    } catch (InvalidProjectIdException $invalidProjectIdException) {
+        $response = new Response('Invalid project ID', []);
```
Comment (norayna) [inline project.php:30]: why don't you use the exception's message as the response message?

Comment (melody-sylvestre) [reply project.php:30]: OK

**Review by norayna — CHANGES_REQUESTED** [review]
1st pass

File: src/Entities/Project.php [hunk 10]
```diff
@@ -36,7 +36,7 @@ private function validateUsersArray(array $users)
     {
         foreach ($users as $user) {
             if (!($user instanceof User)) {
-                throw new InvalidUserArrayDatatypeException("Invalid user array\n");
+                throw new InvalidUserArrayDatatypeException("Input must be instances of User class\n");
```
Comment (norayna) [inline src/Entities/Project.php]: i don't think this is much clearer 😅

File: src/Tests/Entities/ProjectTest.php [hunk 11]
```diff
@@ -86,34 +94,40 @@ public function testSuccessToAssociativeArray()
...
+            'name' => 'name',
+            'client_id' => '1',
+            'deadline' => '21/12/2012',
+            'overdue' => true,
+            'client_name' => 'client name',
+            'client_logo' => 'http://dummyimage.com/200x200.png/ff4444/ffffff',
+            'users' => [
+                ['id' => 1, 'name' => 'name', 'avatar' => 'avatar', 'role' => 'role'],
+                ['id' => 2, 'name' => 'name', 'avatar' => 'avatar', 'role' => 'role'],
+                ['id' => 3, 'name' => 'name', 'avatar' => 'avatar', 'role' => 'role']
```
Comment (norayna) [inline src/Tests/Entities/ProjectTest.php]: i think the user id's are supposed to be strings?

**Review by norayna — CHANGES_REQUESTED** [review]
2nd pass

File: src/Entities/Project.php (hunk 10 above)
Comment (melody-sylvestre) [reply src/Entities/Project.php]: done

File: src/Tests/Entities/ProjectTest.php (hunk 11 above)
Comment (melody-sylvestre) [reply src/Tests/Entities/ProjectTest.php]: fixed it

**Review by norayna — APPROVED** [review]
approved

**Pull request merged**